### PR TITLE
Flag and block users for sending too many DMs

### PR DESCRIPTION
**Pull request by jpaddison3**
After 3 DMs you'll get put up for review. After 10 you'll be blocked from creating any more messages.

image

@darkruby501 curious if this is something y'all are interested in. If desired, I can trivially make the limits a `forumSelect`. I could also make block ~infinite on LW, to give you only half. Or of course, Forum-gate the whole thing.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1203832078942786) by [Unito](https://www.unito.io)

### CODE REVIEW

**Comment by Raemon** [comment]
I've actively wanted something like this. Haven't looked at the implementation yet but definitely seems useful to LW to have to me.

File: packages/lesswrong/server/callbacks/conversationCallbacks.ts
Comment (b0b3rt) [inline packages/lesswrong/server/callbacks/conversationCallbacks.ts:44]: I'd feel a lot better about this if it were broken out into multiple lines that didn't word wrap 😅

**Comment by darkruby501** [comment]
Yeah, I think we want this. Those limits seem fine to me and I wouldn't think we need separate configurability right now. (PS: Ray was recently made czar of New User Moderation, so technically his call. To reduce confusingness on your side though, feel free to direct design/prod questions to me still.)

File: packages/lesswrong/server/callbacks/conversationCallbacks.ts
Comment (b0b3rt) [inline packages/lesswrong/server/callbacks/conversationCallbacks.ts:51]: Should we be handling this by setting `needsReview: true` or via a moderator action?  I think this would work but we wouldn't know why the user was flagged for review, and would need to infer it from seeing that they'd sent DMs to 3 people.

Comment (jpaddison3) [reply packages/lesswrong/server/callbacks/conversationCallbacks.ts:51]: Probably via a moderator action, is what I'm inferring! I don't really know much about it. Tell me more! (or I'll read about it myself)

File: packages/lesswrong/server/callbacks/conversationCallbacks.ts
Comment (jpaddison3) [reply packages/lesswrong/server/callbacks/conversationCallbacks.ts]: Indeed much more readable.

File: packages/lesswrong/server/callbacks/conversationCallbacks.ts
Comment (jpaddison3) [reply packages/lesswrong/server/callbacks/conversationCallbacks.ts]: Now done. Moderator actions are a bit of a mess I'll be honest. Or maybe it's the thing where developers, in general, do not think novel code is good at first blush. Or a combination of both.

File: packages/lesswrong/lib/collections/moderatorActions/collection.ts
Comment (jpaddison3) [inline packages/lesswrong/lib/collections/moderatorActions/collection.ts:15]: @b0b3rt or @Raemon, is this an accurate docstring?

File: packages/lesswrong/server/callbacks/conversationCallbacks.ts
Comment (oetherington) [inline packages/lesswrong/server/callbacks/conversationCallbacks.ts:73]: I don't love `void`-ing these two promises, but maybe I'm being over-paranoid

File: packages/lesswrong/server/callbacks/conversationCallbacks.ts
Comment (oetherington) [inline packages/lesswrong/server/callbacks/conversationCallbacks.ts:10]: This is pretty nitpicky, but I think this comment would be easier to read without the footnote. It could just say "Before a user has been fully approved, keep track of the number of users they've created a conversation with". In general, I feel like footnotes in code comments are better avoided - it feels like a `goto` 😂.

**Review by oetherington — APPROVED** [review]
Nice work - LGTM!

File: packages/lesswrong/server/callbacks/conversationCallbacks.ts
Comment (jpaddison3) [reply packages/lesswrong/server/callbacks/conversationCallbacks.ts:73]: What's the concern? I would hope our mutator system is robust to two mutators changing different fields on the same user. I've spent a lot of time in this part of the code recently and I can't see any reason why not.

In general I think void is a pretty valid tool?

File: packages/lesswrong/server/callbacks/conversationCallbacks.ts
Comment (jpaddison3) [reply packages/lesswrong/server/callbacks/conversationCallbacks.ts:10]: "If you're reading code documentation, you're gonna immediately want to know the caveats."

Pretty reasonable here.

File: packages/lesswrong/lib/collections/moderatorActions/collection.ts
Comment (jpaddison3) [reply packages/lesswrong/lib/collections/moderatorActions/collection.ts:15]: Given that you're on team retreat, will merge and then you can tell me if it's wrong and I'll revisit. (I'm pretty confident in it.)

File: packages/lesswrong/server/callbacks/conversationCallbacks.ts
Comment (oetherington) [reply packages/lesswrong/server/callbacks/conversationCallbacks.ts:71]: I just get a bit worried that if you hypothetically `await`ed a call to this function, then we'd return the to caller in an incomplete state. The function isn't exported though so it should be fine here.

It looks like there's actually no `await` in this function, so it doesn't need to be `async`. That's getting a little nit-picky though.

Basically, I think this is fine to merge - I'm just being difficult 🙃 .

**Pull request merged**
